domaci_maze.py

```diff
@@ -90,495 +77,523 @@ def has_value(self) -> bool:
+    u zavisnoti od potrebe, a u tim funkcijama subelmanove jednacine za racunanje vrednosti.
 
+    Pronalazenje optimalne politike se vrsi pozivom generate optimal_policy, koja vraca politiku
+    kao recnik:
 
-def update_state_value(env: MazeEnvironment, s, v, gamma):
-    rhs = []
-    cell = env.board[s[0], s[1]]
-    for a in env.get_possible_actions(cell, *s):
-        s_new, r, _ = env(s, a)
-        rhs.append(r + gamma * v.get(s_new, 0))  # Default to 0 if s_new not in v
-    return max(rhs) if rhs else 0  # Handle case with no actions
+        dict[position] = optimal_action 
 
+    Takodje se parametrom q_function podesava da li je po V ili Q. Unutar nje se poziva 
+    greedy_action za svako stanje i time dobija politika.
 
-def async_update_all_values(env: MazeEnvironment, v, gamma):
-    """Update values of all states.
+                                    Policy iteration:
 
-    Args:
-        env (MazeEnvironment): The environment to work on.
-        v : Values of other states.
-        gamma : discount factor.
-    """
-    for s in env.get_states():
-        if not env.is_terminal(s):
-            v[s] = update_state_value(env, s, v, gamma)
-    return copy(v)
+        Pozivom funkcije policy_iteration se izvrsava algoritam, opet q_function parametar
+    sluzi za biranje Q ili V funkcije. Unutar nje se izvrsava petlja dok se ne pojave 2 uzastupne
+    iste politike, gde se prvo vrsi estimacija vrednosti V ili Q za zadatu politiku, a zatim se
+    formira greedy politika za te zadate vrednosti i tako do kraja.
 
+'''
 
-def init_values(env):
-    """Randomly initialize states of the given environment."""
-    values = {s: -10 * random() for s in env.get_states()}
 
-    for s in values:
-        if env.is_terminal(s):
-            values[s] = 0
+# v values are dict[position] = value
+def init_v_values(env: MazeEnvironment):
+    return {s.get_position(): -20 * random() if not env.is_terminal(s) else 0 for s in env.get_graph()}
 
-    return values
 
+# q values are dict[position, action] = value
+def init_q_values(env: MazeEnvironment):
+    q = {}
+    for s in env.get_graph():
+        for action in ACTIONS:
+            if env.is_terminal(s):
+                q[s.get_position(), action] = 0
+            else:
+                q[s.get_position(), action] = -20 * random()
+    return q
 
-def draw_values(env, values, ax=None):
-    ax = ax if ax is not None else plt
-    draw_board(env.board, ax=ax)
-    for s in values:
-        ax.text(s[1] - 0.25, s[0] + 0.1, f"{values[s]:.1f}")
 
+# update v value for one node
+def update_v_value(env: MazeEnvironment, position, values, gamma):
+    current_node = env.get_current_pos_node(position)
+    possible_values = []
+    for action in ACTIONS:
+        x = 0
+        for next_node, prob in env.get_action_probabilities(current_node, action):
+            x += prob * (next_node.get_reward() + gamma * values[next_node.get_position()])

```

Comment:
Zar nece values[next_node] uvek biti 0 u pocetku? Odnosno zar ne treba ovo da se radi rekurzivno umesto iterativno. Ili mozda da je graf sortiran topoloski bi i moglo ovako

Reply:
Na pocetku value_iteration funckije su incijalizovane sve te vrednosti funckiom init_v_values/init_q_values

Reply:
pitanje je zasto ovaj algoritam nije realizovan rekurzivno (dinamicko programiranje), da se ne racunaju neke vrednosti vise puta. a i to je naziv lekcije gde je radjen algoritam(?)

Reply:
nije isto ovo dinamicko programiranje i ono sa sausau, ovako radi algoritam ovde

Reply:
aha, znaci samo si ono sa prez implementirao?